hardware/raspberrypi/bootmodes/bootflow_2711.md

```diff
@@ -1,31 +1,74 @@
 
-The boot flow for the Pi4 is as follows:
+The bootflow for the ROM (first stage) is as follows:-
 
 * BCM2711 SoC powers up
-* On board bootrom checks for bootloader recovery file (recovery.bin) on the SD card. If found, it executes it to flash the EEPROM and recovery.bin triggers a reset.
-* Otherwise, the bootrom loads the main bootloader from the EEPROM.
-* Bootloader checks it's inbuilt BOOT_ORDER configuration item to determine what type of boot to do.
-  * SD Card
-  * Network
-  * USB mass storage
+* Read OTP to determine if the `nRPIBOOT` GPIO is configured
+* If `nRPIBOOT` GPIO is high or OTP does NOT define `nRPIBOOT` GPIO 
+   * Check OTP to see if `recovery.bin` can be loaded from SD/EMMC
+      * If SD recovery.bin is enabled then check primary SD/EMMC for `recovery.bin`
+         * Success - run `recovery.bin` and update the SPI EEPROM
+         * Fail - continue
+   * Check SPI EEPROM for second stage loader
+      * Success - run second stage bootloader 
+      * Fail - continue
+* While True
+   * Attempt to load recovery.bin from [USB device boot](../../computemodule/cm-emmc-flashing.md)
+      * Success - run `recovery.bin` and update the SPI EEPROM or switch to USB mass storage device mode
+      * Fail - retry USB device boot
 
+N.B. Currently only CM4 reserves a GPIO for `nRPIBOOT`
 
-## SD Card Boot
-The bootloader loads the files in the [boot folder](../../../configuration/boot_folder.md) according to the [boot options](../../../configuration/config-txt/boot.md) in config.txt
+## recovery.bin
+`recovery.bin` is a minimal second stage program used to reflash the bootloader SPI EEPROM image.
 
-## Network boot
+# Second stage bootloader 
 
-Details of the network booting can be found [here](../bcm2711_bootloader_config.md)
+This section describes the high-level flow of the second stage bootloader.
 
-## USB mass storage boot
-
-USB booting is still under development.
-
-
-## BOOT_ORDER
-
-The `BOOT_ORDER` configuration item is embedded inside the bootloader code in the EEPROM. See the [Pi4 Bootloader Configuration](../bcm2711_bootloader_config.md) page for details on how to change the bootloader configuration.
+Please see the [bootloader configuration](../bcm2711_bootloader_config.md) page for more information about each boot-mode and the [boot folder](../../../configuration/boot_folder.md) page for a description of the GPU firmware files loaded by this stage.
 
+* Initialise clocks and SDRAM
+* Read the EEPROM configuration file
+* Check `PM_RSTS` register to determine if HALT is requested
+   * Check `POWER_OFF_ON_HALT` and `WAKE_ON_GPIO` EEPROM configuration settings.
+   * If `POWER_OFF_ON_HALT` is `1` and `WAKE_ON_GPIO` is `0` then
+      * Use PMIC to power off system
+   * else
+      * While GPIO3 is high OR `WAKE_ON_GPIO` is `1`
```

Comment:
Is the "OR" here correct? It implies that if "`WAKE_ON_GPIO` is `1`" then it'll _never_ exit this sleep-loop?

(as a side-note, https://github.com/raspberrypi/documentation/blob/master/hardware/raspberrypi/bcm2711_bootloader_config.md#configuration-properties doesn't really explain what happens when `POWER_OFF_ON_HALT` is `0` and `WAKE_ON_GPIO` is `0`)